brew-rmtree.rb

```diff
@@ -29,61 +29,67 @@ def bash(command)
     return %x! bash -c #{escaped_command} !
   end
 
-  def rmtree
-    if ARGV.size < 1 or ['-h', '?', '--help'].include? ARGV.first
-      puts USAGE
-      exit 0
-    end
-
-    raise KegUnspecifiedError if ARGV.named.empty?
-
-    if not ARGV.force?
-      ARGV.named.each do |keg_name|
+  def remove_keg(keg_name)
+    # Remove old versions of keg
+    puts bash "brew cleanup #{keg_name}"
 
-        # Remove old versions of keg
-        puts bash "brew cleanup #{keg_name}"
-
-        # Remove current keg
-        puts bash "brew uninstall #{keg_name}"
-
-        deps = bash "join <(brew leaves) <(brew deps #{keg_name})"
-
-        if deps.length > 0
-          puts "Found lingering dependencies"
-          puts deps.chomp
+    # Remove current keg
+    puts bash "brew uninstall #{keg_name}"
+  end
 
-          deps = deps.split("\n")
+  def deps(keg_name)
+    deps = bash "join <(brew leaves) <(brew deps #{keg_name})"
+    deps.split("\n")
+  end
 
-          deps.each do |dep|
-            dep = dep.chomp
+  def reverse_deps(keg_name)
+    reverse_deps = bash "brew uses --installed #{keg_name}"
+    reverse_deps.split("\n")
+  end
 
-            if !dep.empty?
-              # Check if anything currently installed uses the dependency
-              dep_deps = bash "brew uses --installed #{dep}"
-              dep_deps = dep_deps.chomp
+  def rmtree(keg_name)
+    # Check if anything currently installed uses the keg
+    reverse_deps = reverse_deps(keg_name)
 
-              if dep_deps.length > 0
-                puts "Not removing dependency #{dep} because other installed packages depend on it:"
-                puts dep_deps
-              else
-                # Nothing claims to depend on it
-                puts "Removing dependency #{dep}..."
-                puts bash "brew rmtree #{dep}"
-              end
-            end
-          end
-        else
-          puts "No dependencies left on system for #{keg_name}."
+    if reverse_deps.length > 0
+      puts "Not removing #{keg_name} because other installed kegs depend on it:"
+      puts reverse_deps.join("\n")
+    else
+      # Nothing claims to depend on it
+      puts "Removing #{keg_name}..."
+      remove_keg(keg_name)
+
+      deps = deps(keg_name)
+      if deps.length > 0
+        puts "Found lingering dependencies:"
+        puts deps.join("\n")
+
+        deps.each do |dep|
+          puts "Removing dependency #{dep}..."
+          rmtree dep
         end
+      else
+        puts "No dependencies left on system for #{keg_name}."
       end
-    else
+    end
+  end
+
+  def main
+    if ARGV.size < 1 or ['-h', '?', '--help'].include? ARGV.first
+      puts USAGE
+      exit 0
+    end
+
+    if ARGV.force?
       puts "--force is not supported."
+      exit 1
     end
-  rescue MultipleVersionsInstalledError => e
-    ofail e
-    puts "Use `brew rmtree --force #{e.name}` to remove all versions."
```

Comment:
Hm, yes, this rescue block...

Your first point is perfectly valid and is completely on me for not implementing it, I got lazy and didn't think it would ever be used.

And your second point, I was going to say I didn't agree with, but that would have been premature. The rescue would have been necessary if I was using `ARGV.kegs` to iterate over, and was in fact doing so before publishing. But I'm not. I'm using `named` which throws no such exception so you would be correct in removing it. Thanks for the find.